### PR TITLE
feat: 增加 DB Listen Port 映射规则

### PR DESCRIPTION
**Pull request by fit2bot**
feat: 增加 DB Listen Port 映射规则

### CODE REVIEW

File: apps/applications/api/application.py [hunk 1]
```diff
@@ -37,3 +41,30 @@ def get_tree(self, request, *args, **kwargs):
...
+        'db_info': 'applications.view_application',
+    }
+
+    http_method_names = ['get', 'post']
+
+    def list(self, request, *args, **kwargs):
+        ports = db_port_manager.get_already_use_ports()
+        return Response(data=ports, status=status.HTTP_200_OK)
+
+    @action(methods=['post'], detail=False, url_path='db-info')
```
Comment (ibuler) [inline apps/applications/api/application.py:59]: action: GET

File: apps/applications/utils/db_port_mapper.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            if not db:
+                msg = 'Database not exists, database id: {}'.format(db_id)
+            else:
+                msg = ''
+        else:
+            db = None
+            msg = 'Port not in port-db mapper, port: {}'.format(port)
+        return db, msg
+
+    def get_next_usable_port(self):
```
Comment (ibuler) [inline apps/applications/utils/db_port_mapper.py:76]: avaiable

File: apps/applications/utils/db_port_mapper.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    """ 管理端口-数据库ID的映射, Magnus 要使用 """
+
+    CACHE_KEY = 'PORT_DB_MAPPER'
+
+    def __init__(self):
+        self.port_start = settings.MAGNUS_DB_PORTS_START
+        self.port_limit = settings.MAGNUS_DB_PORTS_LIMIT_COUNT
+        self.port_end = self.port_start + self.port_limit
+        # 可以使用的端口列表
+        self.all_usable_ports = [i for i in range(self.port_start, self.port_end+1)]
```
Comment (ibuler) [inline apps/applications/utils/db_port_mapper.py:25]: list(range(self.port_start, self.port_end+1))

File: apps/applications/utils/db_port_mapper.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    @property
+    def magnus_listen_port_range(self):
+        return f'{self.port_start}-{self.port_end}'
+
+    def init(self):
+        db_ids = Application.objects.filter(category=AppCategory.db).values_list('id', flat=True)
```
Comment (ibuler) [inline apps/applications/utils/db_port_mapper.py:32]: tmp_to_root

File: apps/applications/utils/db_port_mapper.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            len(self.all_usable_ports), len(already_use_ports)
+        )
+        logger.warning(msg)
+
+    def get_already_use_ports(self):
+        mapper = self.get_mapper()
+        return sorted(list(mapper.keys()))
+
+    def get_mapper(self):
+        return cache.get(self.CACHE_KEY, {})
```
Comment (ibuler) [inline apps/applications/utils/db_port_mapper.py:94]: reinit

**Comment by BaiJiangJie** [comment]
image

**Comment by BaiJiangJie** [comment]
image

**Comment by BaiJiangJie** [comment]
image

**Comment by BaiJiangJie** [comment]
image

**Review by BaiJiangJie — APPROVED** [review]

**Pull request merged**